kong/db/dao/plugins.lua

```diff
@@ -22,6 +21,99 @@ local ngx_WARN = ngx.WARN
 local ngx_DEBUG = ngx.DEBUG
 
 
+local function has_at_least_one_common_protocol_with_route(plugin, route)
+  local plugin_prot = plugin.protocols
+  local route_prot = route.protocols
+  -- plugin.protocols and route.protocols are both sets provided by the schema
+  -- this means that they can be iterated as over an array, and queried as a hash
+  for i = 1, #plugin_prot do
+    if route_prot[plugin_prot[i]] then
+      return true
+    end
+  end
+end
+
+
+local function has_at_least_one_compatible_route_in_service(self, plugin, service_pk)
+  local routes, err_t, offset
+  repeat
+    routes, err_t, offset = self.db.routes:page_for_service(service_pk, nil, offset)
```

Comment:
any particular reason for not using `each_for_service` with a `for` loop instead? I saw the repeat/until and had to do a double-take

Reply:
Because I'm a dork, that's why :P . Rewrote using `each_for_service`